### ISSUE TITLE
Implement TestGiraffaCLI

### ISSUE THREAD

**Issue by shvachko** [GiraffaFS/giraffa#15]
Original [issue 15](https://code.google.com/p/giraffa/issues/detail?id=15) created by shvachko on 2012-08-17T05:08:22.000Z:

Currently we only test very simplistic cases.

We should make some more interesting / intensive tests. For example, something to that creates many entries in the Namespace, or an IO test for Giraffa.

**Comment by shvachko** [comment]
Comment [#1](https://code.google.com/p/giraffa/issues/detail?id=15#c1) originally posted by shvachko on 2012-08-21T18:50:47.000Z:

My proposal on this is to adopt in some form TestHDFSCLI.
This test defines what behavior is expected from different HDFS commands.
We should of course comply with most of them. And we can add specific for Giraffa test cases in the same format as TestHDFSCLI.
We can try to start with just overloading TestHDFSCLI, defining Giraffa specific setUp() and tearDown(), and see how it holds.

**Comment by shvachko** [comment]
Comment [#2](https://code.google.com/p/giraffa/issues/detail?id=15#c2) originally posted by shvachko on 2012-09-12T22:37:54.000Z:

I have actually started looking into this. It is difficult to figure out how to start -- it looks like we should copy TestHDFSCLI and make a test that extends CLITestHelper. From there, can we use HBaseMiniCluster? Should we create racks and hosts like they do? Basically I am a little confused on how to get us started on this.

**Comment by shvachko** [comment]
Comment [#3](https://code.google.com/p/giraffa/issues/detail?id=15#c3) originally posted by shvachko on 2012-09-13T02:11:01.000Z:

We should derive TestGiraffaCLI from TestHDFSCLI. We than can override beforeClass() and afterClass() to start / stop Giraffa cluster. We also need to provide client configuration so that GiraffaFileSystem instantiated instead of HDFS. Does it make sense?

**Comment by shvachko** [comment]
Comment [#4](https://code.google.com/p/giraffa/issues/detail?id=15#c4) originally posted by shvachko on 2012-09-13T02:25:15.000Z:

At the high level, it makes sense -- I guess I'm just concerned about how similar it needs to be to TestHDFSCLI on the code level. Well, I will do my best to come up with a patch and then I'll post it here.

**Comment by shvachko** [comment]
Comment [#5](https://code.google.com/p/giraffa/issues/detail?id=15#c5) originally posted by shvachko on 2012-09-13T05:19:02.000Z:

At the high level it should be

class TestGiraffaCLI extends TestHDFSCLI {
  @Override
  setUp() {
    // your code here
  }
  @Override
  tearDown() {
    // your code here
  }
}

Then running TestGiraffaCLI will call all test cases of TestHDFSCLI.
We just need to run it on mini Giraffa cluster rather than HDFS.
In order to do that we should define our own setUp() and tearDown(), which @Override the ones of TestHDFSCLI. So this isn't a lot of new code. ALL test cases are reused from TestHDFSCLI.
This is on the high level. But some tests may not work for Giraffa.

**Comment by shvachko** [comment]
Comment [#6](https://code.google.com/p/giraffa/issues/detail?id=15#c6) originally posted by shvachko on 2012-09-14T15:24:16.000Z:

Changing title to reflect work.

**Comment by shvachko** [comment]
Comment [#7](https://code.google.com/p/giraffa/issues/detail?id=15#c7) originally posted by shvachko on 2012-09-16T20:20:53.000Z:

This is what I have been working on. I know I am using CLITestHelper here, I will change to TestHDFSCLI once I am sure I've gotten it working. Currently I am having issues getting the test to recognize Giraffa. It seems no matter what it tries to contact HDFS and then it crashes.

**Comment by shvachko** [comment]
Comment [#8](https://code.google.com/p/giraffa/issues/detail?id=15#c8) originally posted by shvachko on 2012-09-16T20:21:59.000Z:

testGRFAConf.xml is just a renamed testHDFSConf.xml file from the original test.

**Comment by shvachko** [comment]
Comment [#9](https://code.google.com/p/giraffa/issues/detail?id=15#c9) originally posted by shvachko on 2012-09-18T15:51:04.000Z:

Getting along a lot farther now. Here are my latest changes. Now running into Exceptions during "testAll". Let me know if I should remove tests that cause an issue or just how to proceed in general. The current issue is DFSADMIN.

**Comment by shvachko** [comment]
Comment [#10](https://code.google.com/p/giraffa/issues/detail?id=15#c10) originally posted by shvachko on 2012-09-19T04:36:43.000Z:

Okay I can get the tests to run, however they all get skipped. I am not sure what to do about the execute() and expandCommand() functions. I feel like I'm missing something about these tests because I don't see any of them actually running commands on Giraffa. Do I need to have an actual cluster running? Is the MiniCluster not enough? I feel like it is not recognizing grfa:// as the "tag".

Look at my latest work.testGRFAConf.xml is unchanged.

**Comment by shvachko** [comment]
Comment [#11](https://code.google.com/p/giraffa/issues/detail?id=15#c11) originally posted by shvachko on 2012-09-19T07:01:53.000Z:

Okay got it working. We are currently passing 18 tests. Yay! In order to get this working I had to revert it to "extends CLITestHelper". The reason is because TestHDFSCLI overrides tearDown(), and CLITestHelper only outputs results from that tearDown() call. TestHDFSCLI's tearDown() tries to close a DistributedFileSystem variable that TestGRFACLI does not define (so its null, so we get a NPE). Therefore switching to CLITestHelper allows us to properly get the results.

Here is my current TestGRFACLI.java file and the tailed output after I ran it.

There were many tests that seemed to create a file with 0 blocks, leading to ArrayIndexOutOfBoundsException in getBlockLocations().

Also I think we need to be careful with how this test is running. I have random files being written and deleted all over the place.

**Comment by shvachko** [comment]
Comment [#14](https://code.google.com/p/giraffa/issues/detail?id=15#c14) originally posted by shvachko on 2012-09-21T03:06:14.000Z:

With the following changes, it now passes 30% of the tests!

The main issues found by trying to conform to the expected RegexComparator.

The first 2 discovered issues: 
1) Files by default are given permissions with the executable bits on. HDFS expects them to be off by default creation of a file; currently we use what is passed by function.
2) We need to grab the shortUserName from the UGI; currently we use what is passed by function.
3) We need to set the default usergroup from the UGI to be "supergroup". This is what HDFS expects according to testHDFSCLI.xml.

**Comment by shvachko** [comment]
Comment [#16](https://code.google.com/p/giraffa/issues/detail?id=15#c16) originally posted by shvachko on 2012-09-21T05:55:30.000Z:

Simply amazing. With some further changes to testGRFAConf.xml (just replacing hardcoded "hdfs" with "grfa") we are now passing 40% of the tests!

At some point during the tests the cluster breaks down though and it takes about 10 minutes for it to recover and continue the tests. I can look into which test is causing this. This was a lot for tonight though, we are almost half way there with minimal effort! Woot!

Change and output attached.

**Comment by shvachko** [comment]
Comment [#17](https://code.google.com/p/giraffa/issues/detail?id=15#c17) originally posted by shvachko on 2012-09-21T06:20:18.000Z:

The latest issue appears to be code related. Whenever we do a du or an ls it looks like the Regex is expecting the output to have a URI prefix, so like grfa:///dir0/, instead of just /dir0. You can check this by looking at the result for Test ID: [18]. Look at what the Regex wants, and what the actual output is. This sort of issue plagues about ~30 other failing tests.

**Comment by shvachko** [comment]
Comment [#18](https://code.google.com/p/giraffa/issues/detail?id=15#c18) originally posted by shvachko on 2012-09-21T08:32:12.000Z:
1. I think you are on the right track with extending CLITestHelper rather than TestHDFSCLI.
2. The main point is that we should not change a single byte in testHDFSConf.xml. And should not copy and rename it to guarantee we are 100% HDFS compatible.
3. Yes you are right the problem is that hdfs:/ is explicitly written in testHDFSConf.xml, even though there is NAMESPACE variable, which should be used instead, and then replaced with the proper schema. I see there was an attempt to make TestHDFSCLI work for hdfs: and file: schemes, but now this path is abandoned. If you feel we should fix it we can file HDFS jira and fix it there. Then Giraffa should be very easy.
4. For now. I see that you correctly replace hdfs:/ with grfa:/ in expandCommand(). The problem I think with comparators that have hdfs:/ hardcoded in testHDFSConf.xml. How we can modify comparators? not in the .xml file, but programmatically in the test?
5. I occasionally see tabs in your code. You should probably check your eclipse preferences.

**Comment by shvachko** [comment]
Comment [#19](https://code.google.com/p/giraffa/issues/detail?id=15#c19) originally posted by shvachko on 2012-09-21T14:38:04.000Z:
1. Awesome. :)
2. I agree -- infact testGRFAConf.xml and testHDFSConf.xml are currently 1:1, I am only using testGRFAConf.xml so I can kind of tweak around if I need to. Indeed the ultimate goal is to rely soly on testHDFSConf.xml.
3. Yeah there are definitely some hard-coded dependencies; I am unsure whether to file a JIRA or not because for HDFS it might make sense. Their test is testHDFSConf, right? So what is wrong for them to have some hardcoded dependencies if the test is meant solely for HDFS? The JIRA would also have to be filed against 0.22.0; I don't know that filing one for 1.0 or 2.0.0 would help; they may have changed the code even further.
4. Yeah in expandCommand() I handle a lot of the issues; yes the issue is in the Comparators. They are generated via the file as part of CLITestHelper. The only thing I can think of right away is to have a small Java code segment in setUp() make a copy of testHDFSConf.xml and just replace all "hdfs" with "grfa". I don't think the option is available for us in the test.
5. Those dang tabs.

**Comment by shvachko** [comment]
Comment [#20](https://code.google.com/p/giraffa/issues/detail?id=15#c20) originally posted by shvachko on 2012-09-21T18:11:56.000Z:

&lt;empty&gt;

**Comment by shvachko** [comment]
Comment [#22](https://code.google.com/p/giraffa/issues/detail?id=15#c22) originally posted by shvachko on 2012-09-21T18:44:38.000Z:

Here is a run down of the current issue: grfa://localhost:9000/ comes from renaming NAMENODE in expandCommand(), having it be just grfa:/// works too, but either way it fails the comparator which expects grfa://<host>:<port>/. Why does ours not output host and port too?

12/09/21 11:32:44 INFO cli.CLITestHelper: -------------------------------------------
12/09/21 11:32:44 INFO cli.CLITestHelper:                     Test ID: [18]
12/09/21 11:32:44 INFO cli.CLITestHelper:            Test Description: [ls:  non-existent file/directory in Namenode's path ]
12/09/21 11:32:44 INFO cli.CLITestHelper: 
12/09/21 11:32:44 INFO cli.CLITestHelper:               Test Commands: [-fs grfa://localhost:9000/ -ls grfa://localhost:9000//file1]
12/09/21 11:32:44 INFO cli.CLITestHelper: 
12/09/21 11:32:44 INFO cli.CLITestHelper: 
12/09/21 11:32:44 INFO cli.CLITestHelper:                  Comparator: [RegexpComparator]
12/09/21 11:32:44 INFO cli.CLITestHelper:          Comparision result:   [fail]
12/09/21 11:32:44 INFO cli.CLITestHelper:             Expected output:   [^ls: Cannot access grfa://\w+[-.a-z0-9]*:[0-9]+/file1: No such file or directory.]
12/09/21 11:32:44 INFO cli.CLITestHelper:               Actual output:   [ls: Cannot access grfa:////file1: No such file or directory.
]

**Comment by shvachko** [comment]
Comment [#23](https://code.google.com/p/giraffa/issues/detail?id=15#c23) originally posted by shvachko on 2012-09-22T07:21:43.000Z:

Solved that issue. The URI needs to be set to something like grfa://localhost:port in the GiraffaTestUtils and in the giraffa-site.xml and replacing "NAMENODE" in the commands.

The next issue is that -du returns the results without the URI prefix'd. Not sure how to handle this issue; I think the testHDFSConf.xml file needs to be changed this time -- Konstantin do you know what might be causing this?

12/09/21 23:46:18 INFO cli.CLITestHelper: -------------------------------------------
12/09/21 23:46:18 INFO cli.CLITestHelper:                     Test ID: [31]
12/09/21 23:46:18 INFO cli.CLITestHelper:            Test Description: [du: file using absolute path]
12/09/21 23:46:18 INFO cli.CLITestHelper: 
12/09/21 23:46:18 INFO cli.CLITestHelper:               Test Commands: [-fs grfa://localhost:9000 -put file:/home/hadoop/git/giraffa/build/test/cache//data15bytes /data15bytes]
12/09/21 23:46:18 INFO cli.CLITestHelper:               Test Commands: [-fs grfa://localhost:9000 -du /data15bytes]
12/09/21 23:46:18 INFO cli.CLITestHelper: 
12/09/21 23:46:18 INFO cli.CLITestHelper:            Cleanup Commands: [-fs grfa://localhost:9000 -rm /data15bytes]
12/09/21 23:46:18 INFO cli.CLITestHelper: 
12/09/21 23:46:18 INFO cli.CLITestHelper:                  Comparator: [TokenComparator]
12/09/21 23:46:18 INFO cli.CLITestHelper:          Comparision result:   [pass]
12/09/21 23:46:18 INFO cli.CLITestHelper:             Expected output:   [Found 1 items]
12/09/21 23:46:18 INFO cli.CLITestHelper:               Actual output:   [Found 1 items
15  /data15bytes
]
12/09/21 23:46:18 INFO cli.CLITestHelper:                  Comparator: [RegexpComparator]
12/09/21 23:46:18 INFO cli.CLITestHelper:          Comparision result:   [fail]
12/09/21 23:46:18 INFO cli.CLITestHelper:             Expected output:   [^15( |\t)_grfa://\w+[-.a-z0-9]_:[0-9]*/data15bytes]
12/09/21 23:46:18 INFO cli.CLITestHelper:               Actual output:   [Found 1 items
15  /data15bytes
]
12/09/21 23:46:18 INFO cli.CLITestHelper:

**Comment by shvachko** [comment]
Comment [#24](https://code.google.com/p/giraffa/issues/detail?id=15#c24) originally posted by shvachko on 2012-09-23T10:06:33.000Z:

Found the issue -- we are not storing the full URI path in the INodes. We are storing just the regular path based on the RowKey. This is incorrect; we need to make sure INode's path variable resolves to a qualified URI, not a simple Path.

Here is the work I've done to implement it and the latest output. In summary:

12/09/23 02:57:50 INFO cli.CLITestHelper: Summary results:
12/09/23 02:57:50 INFO cli.CLITestHelper: ----------------------------------

12/09/23 02:57:50 INFO cli.CLITestHelper:                Testing mode: test
12/09/23 02:57:50 INFO cli.CLITestHelper: 
12/09/23 02:57:50 INFO cli.CLITestHelper:              Overall result: --- FAIL ---
12/09/23 02:57:50 INFO cli.CLITestHelper:                # Tests pass: 405 (64%)
12/09/23 02:57:50 INFO cli.CLITestHelper:                # Tests fail: 224 (35%)
12/09/23 02:57:50 INFO cli.CLITestHelper:          # Validations done: 1832 (each test may do multiple validations)
12/09/23 02:57:50 INFO cli.CLITestHelper: 

We are now passing 64% of the tests! The only things that aren't passing are commands that aren't fully implemented yet.

**Comment by shvachko** [comment]
Comment [#26](https://code.google.com/p/giraffa/issues/detail?id=15#c26) originally posted by shvachko on 2012-09-24T01:09:13.000Z:

&lt;empty&gt;

**Comment by shvachko** [comment]
Comment [#27](https://code.google.com/p/giraffa/issues/detail?id=15#c27) originally posted by shvachko on 2012-09-24T05:45:56.000Z:

More improvements were made. Now passing 66% of tests! I've opened up Issue #&nbsp;40 to describe the new problem I had to fix.

I removed the need to altar giraffa-site.xml or the TestUtils.

Attaching latest changes and outputs.

**Comment by shvachko** [comment]
Comment [#28](https://code.google.com/p/giraffa/issues/detail?id=15#c28) originally posted by shvachko on 2012-09-24T05:51:45.000Z:

&lt;empty&gt;

**Comment by shvachko** [comment]
Comment [#29](https://code.google.com/p/giraffa/issues/detail?id=15#c29) originally posted by shvachko on 2012-09-24T05:54:18.000Z:

The test only takes 120 seconds to run now. Major improvement!

**Comment by shvachko** [comment]
Comment [#30](https://code.google.com/p/giraffa/issues/detail?id=15#c30) originally posted by shvachko on 2012-09-24T17:47:30.000Z:

I think the correct methodology here is to have the GiraffaFileSystem turn all source paths into URI-prefixed paths using MakeQualified(). When the URI-prefixed path makes it to NamespaceProcessor, we will strip out the URI for the sake of generating RowKeys and INodes, but still retain the URI-prefixed path in the INode's path variable (not the RowKey's path variable though).

Otherwise, the next fix would be to convert the source paths into URIs at the NamespaceProcessor layer, but that requires access to the GiraffaConfiguration held by GiraffaFileSystem, and makes it much harder to implement.

I will post up a working patch tonight which I believe will be the FINAL patch -- meaning it will encapsulate ALL the TestHDFSCLI tests that we can possible pass using the current state of the GiraffaFileSystem.

**Comment by shvachko** [comment]
Comment [#31](https://code.google.com/p/giraffa/issues/detail?id=15#c31) originally posted by shvachko on 2012-09-25T06:07:45.000Z:

I am almost there -- passing percentage dropped back down to 66%. I will look into it. Here is my patch with the methodology from the previous post.

Output using this patch is attached as well as the unchanged TestGiraffaCLI.

**Comment by shvachko** [comment]
Comment [#32](https://code.google.com/p/giraffa/issues/detail?id=15#c32) originally posted by shvachko on 2012-09-25T06:41:25.000Z:

Some comments:
1. TEST_FILES_DIR should point to something "build/...", but should never generate test files in "src/..."
2. TEST_CONFIG_FILE. I don't have testHDFSConf.xml in src/test. You probably copied it manually from hdfs. Is there way to refer to that file directly in hdfs source jar, rather than copying it to giraffa?

Let's first commit issue(s) which fix bugs detected by the test, then the test itself and disable it.

**Comment by shvachko** [comment]
Comment [#33](https://code.google.com/p/giraffa/issues/detail?id=15#c33) originally posted by shvachko on 2012-09-25T06:56:37.000Z:

Can you also make a patch rather than attaching java file.

**Comment by shvachko** [comment]
Comment [#34](https://code.google.com/p/giraffa/issues/detail?id=15#c34) originally posted by shvachko on 2012-09-25T06:57:39.000Z:

TEST_FILES_DIR is an internal variable that just points to some of the data we need to have ourselves.

TEST_CONFIG_FILE points to testHDFSConf.xml -- there is no way to pull it out of the source jar because we never pull the source jar from Ivy. Should we pull the source jars?

**Comment by shvachko** [comment]
Comment [#35](https://code.google.com/p/giraffa/issues/detail?id=15#c35) originally posted by shvachko on 2012-09-25T07:04:23.000Z:

I was attaching just the java file out of simplicity because the test was still in the works. Here is the current "patch" I am using to do the test -- keep in mind you will still need TestHDFSConf.xml and the cli_testdata directory from the HDFS tests.

**Comment by shvachko** [comment]
Comment [#36](https://code.google.com/p/giraffa/issues/detail?id=15#c36) originally posted by shvachko on 2012-09-25T07:09:30.000Z:

This patch (15.6) incorporates the fixes from Issue #&nbsp;39 and Issue #&nbsp;40. I am working on generating a separate patch for them right now.

**Comment by shvachko** [comment]
Comment [#37](https://code.google.com/p/giraffa/issues/detail?id=15#c37) originally posted by shvachko on 2012-09-25T07:47:17.000Z:

&lt;empty&gt;

**Comment by shvachko** [comment]
Comment [#38](https://code.google.com/p/giraffa/issues/detail?id=15#c38) originally posted by shvachko on 2012-09-26T07:05:30.000Z:

> TEST_FILES_DIR is an internal variable that just points to some of the data we need to have ourselves.

I agree, but this data should not be generated in src/. It should be generated in build/. That way it will be properly cleaned up by clean target.

I checked the jars. Seems like the xml file is not there. May be we should add a compatibility package and add those things there.

**Comment by shvachko** [comment]
Comment [#39](https://code.google.com/p/giraffa/issues/detail?id=15#c39) originally posted by shvachko on 2012-09-26T07:16:42.000Z:

What do you mean by "add a compatibility package"?

**Comment by shvachko** [comment]
Comment [#40](https://code.google.com/p/giraffa/issues/detail?id=15#c40) originally posted by shvachko on 2012-09-26T07:37:09.000Z:

May be not a package but rather src/test/compatibility
Then we can have a target to run these tests.

**Comment by shvachko** [comment]
Comment [#41](https://code.google.com/p/giraffa/issues/detail?id=15#c41) originally posted by shvachko on 2012-09-26T18:30:44.000Z:

Oh so you mean like a separate pile of tests for compatibility. I'll approve of that -- we don't even need to "disable" the test then; we will just not tell Jenkins to run the compatibility tests; we will run them on our local dev boxes instead.

I think after Issue 39 is cleared up, we will be on a good path to completing this. I would like to see all 8 of the cat tests pass as well though. :(

**Comment by shvachko** [comment]
Comment [#42](https://code.google.com/p/giraffa/issues/detail?id=15#c42) originally posted by shvachko on 2012-09-28T16:03:23.000Z:

Konstantin, the other failing tests are DFSAdmin tests -- if you follow the code they are created at the execute() method by CmdFactoryDFS in testHDFSCLI. They create a DFSAdmin() object that specifically relies on a DistributedFileSystem to be created. I may need to modify or override this class to do the test. Maybe create GRFAAdmin or something.

**Comment by shvachko** [comment]
Comment [#43](https://code.google.com/p/giraffa/issues/detail?id=15#c43) originally posted by shvachko on 2012-09-28T17:03:46.000Z:

I've created a separate issue, issue 43, to discuss how to fix the DFSAdmin tests. We should move our discussion for this new issue there. I believe issue 39, 40, and 41 are now more or less solved. Fixing DFSAdmin will get us to ~90% passing, I believe.

**Comment by shvachko** [comment]
Comment [#44](https://code.google.com/p/giraffa/issues/detail?id=15#c44) originally posted by shvachko on 2012-09-28T17:07:29.000Z:

&lt;empty&gt;

**Comment by shvachko** [comment]
Comment [#45](https://code.google.com/p/giraffa/issues/detail?id=15#c45) originally posted by shvachko on 2012-09-29T02:50:19.000Z:

&lt;empty&gt;

**Comment by shvachko** [comment]
Comment [#46](https://code.google.com/p/giraffa/issues/detail?id=15#c46) originally posted by shvachko on 2012-09-29T17:11:18.000Z:

&lt;empty&gt;

**Comment by shvachko** [comment]
Comment [#47](https://code.google.com/p/giraffa/issues/detail?id=15#c47) originally posted by shvachko on 2012-10-24T05:01:31.000Z:

we really need to finish this.
Can we add some verification that if 75% (or better the exact number) of tests pass then TestGiraffaCLI is successful. I think there should be a way. We will move the threshold higher as the features are implemented.

**Comment by shvachko** [comment]
Comment [#48](https://code.google.com/p/giraffa/issues/detail?id=15#c48) originally posted by shvachko on 2012-10-24T07:56:04.000Z:

This is what I am using so far for running this.
These are my results:
12/10/24 00:53:05 INFO cli.CLITestHelper:                Testing mode: test
12/10/24 00:53:05 INFO cli.CLITestHelper: 
12/10/24 00:53:05 INFO cli.CLITestHelper:              Overall result: --- FAIL ---
12/10/24 00:53:05 INFO cli.CLITestHelper:                # Tests pass: 486 (77%)
12/10/24 00:53:05 INFO cli.CLITestHelper:                # Tests fail: 143 (22%)

I am not sure of how to change the overall result.

**Comment by shvachko** [comment]
Comment [#49](https://code.google.com/p/giraffa/issues/detail?id=15#c49) originally posted by shvachko on 2012-10-24T15:46:42.000Z:

I think maybe we can add verification at the testAll() layer and remove the fail() lines but that would require copy pasting a lot of code. I can try to do it and post the code up here afterwards.

**Comment by shvachko** [comment]
Comment [#51](https://code.google.com/p/giraffa/issues/detail?id=15#c51) originally posted by shvachko on 2012-10-30T06:52:09.000Z:

GiraffaAdmin belongs to a different issue. Why are you adding it here?

**Comment by shvachko** [comment]
Comment [#52](https://code.google.com/p/giraffa/issues/detail?id=15#c52) originally posted by shvachko on 2012-10-30T07:21:45.000Z:

Here is an idea. In TestGiraffaCLI.tearDown() we catch the AssertError. If the message says "One of the tests failed. " we calculate number of failed tests and then assert if it is not high enough.

try {
  super.tearDown(); // displayResults
} catch(AssertionError a) {
  if(!a.getLocalizedMessage().startsWith("One of the tests failed."))
    throw a;
}
// calculate # failed tests, and assert if more than the threshold

**Comment by shvachko** [comment]
Comment [#53](https://code.google.com/p/giraffa/issues/detail?id=15#c53) originally posted by shvachko on 2012-10-30T07:31:20.000Z:

And let's not block it on admin issue.

**Comment by shvachko** [comment]
Comment [#54](https://code.google.com/p/giraffa/issues/detail?id=15#c54) originally posted by shvachko on 2012-10-30T18:08:57.000Z:

I think it SHOULD block on the Admin issue because if we don't use GiraffaAdmin in TestGiraffaCLI then we are losing all the count and qouta passing tests.

I included GiraffaAdmin in this for free; using DFSAdmin is strange when we are using a different FS.

I think we can go ahead with your approach; I'll update it when I have some free time.

**Comment by shvachko** [comment]
Comment [#55](https://code.google.com/p/giraffa/issues/detail?id=15#c55) originally posted by shvachko on 2012-11-10T18:57:11.000Z:

Here is a patch for TestGiraffaCLI. I added and tested the AssertionError catch. It works. I set the standard for 77% passing rate.

The current patch relies on issue 43 being implemented. So perhaps we should commit issue 43 first, and then 15.

Check it out.

**Comment by shvachko** [comment]
Comment [#56](https://code.google.com/p/giraffa/issues/detail?id=15#c56) originally posted by shvachko on 2012-11-10T18:59:00.000Z:

I think we need to still figure out a way to properly include the test data into our test suite though. Would be nice if we could get it from our lib/hadoop*-test.jar or something.

**Comment by shvachko** [comment]
Comment [#57](https://code.google.com/p/giraffa/issues/detail?id=15#c57) originally posted by shvachko on 2012-11-15T05:52:46.000Z:

The patch looks good. We should commit it once you figure out how to get the test data and config file from that jar. Or add those to your patch.

**Comment by shvachko** [comment]
Comment [#58](https://code.google.com/p/giraffa/issues/detail?id=15#c58) originally posted by shvachko on 2012-11-15T06:07:23.000Z:

Oh. One nit. 77 should be a constant or better a parameter to TestGiraffaCLI.

**Comment by shvachko** [comment]
Comment [#60](https://code.google.com/p/giraffa/issues/detail?id=15#c60) originally posted by shvachko on 2012-11-16T20:00:56.000Z:

I think for now we should include them in the src/test directory; we can remove them following the 0.23.4 upgrade (issue 55).

I'll post up a better patch shortly.

**Comment by shvachko** [comment]
Comment [#61](https://code.google.com/p/giraffa/issues/detail?id=15#c61) originally posted by shvachko on 2012-11-16T21:04:30.000Z:

This is an all-inclusive patch. It contains everything -- including the test-data and GiraffaAdmin (issue 43).

I've separated the patches out however, for your ease. I know the Issue43 patch doesn't belong here, but it is tightly coupled at the moment.

I've fixed your nit as well as some other outstanding issues. Check out 15.5 first.
